app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.lockbox.store
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.os.SystemClock
+import android.preference.PreferenceManager
+import io.reactivex.Observable
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.rxkotlin.addTo
+import io.reactivex.subjects.ReplaySubject
+import io.reactivex.subjects.Subject
+import mozilla.lockbox.action.LifecycleAction
+import mozilla.lockbox.flux.Dispatcher
+import mozilla.lockbox.support.Constant
+
+class AutoLockStore(
+    val dispatcher: Dispatcher = Dispatcher.shared,
+    val settingStore: SettingStore = SettingStore.shared,
+    val lifecycleStore: LifecycleStore = LifecycleStore.shared
+) : ContextStore {
+
+    companion object {
+        val shared by lazy { AutoLockStore() }
+    }
+
+    private val compositeDisposable = CompositeDisposable()
+    private lateinit var preferences: SharedPreferences
+
+    val lockRequired: Observable<Boolean> = ReplaySubject.createWithSize(1)
+
+    init {
+        lifecycleStore.lifecycleFilter
+            .filter { it == LifecycleAction.Foreground }
+            .subscribe {
+                checkElapsedTime()
+            }
+            .addTo(compositeDisposable)
+
+        lifecycleStore.lifecycleFilter
+            .filter { it == LifecycleAction.Background }
+            .switchMap { settingStore.autoLockTime.take(1) }
+            .subscribe {
+                val currentSystemTime = SystemClock.elapsedRealtime()
+                preferences
+                    .edit()
+                    .putLong(Constant.Key.autoLockTimerDate, currentSystemTime + it.ms)
+                    .apply()
+            }
+            .addTo(compositeDisposable)
+    }
+
+    override fun injectContext(context: Context) {
+        preferences = PreferenceManager.getDefaultSharedPreferences(context)
+    }
+
+    private fun checkElapsedTime() {
+        val autoLockTimerDate = preferences.getLong(Constant.Key.autoLockTimerDate, -1)
+        val currentSystemTime = SystemClock.elapsedRealtime()
+
+        // edge case: someone has backgrounded the app after the device has been running for x minutes.
+        // they reboot the device and leave it running for x minutes + some time less than autolocktime setting
+        // app will be unlocked :(
+        val diff = autoLockTimerDate - currentSystemTime
```

Comment:
This is a little concerning to me, but not sure if we can properly address it.

I wonder if adding a broadcast receiver for system boot is worth exploring.  My initial thought is to reset the `autolockTimerDate` to 0 (or delete it altogether).  I _think_ that would lock by default.

Thoughts? (cc @jhugman)

Reply:
I've done a little investigating; we can either implement a broadcast receiver, or (IMO best solution), store the value found at `/proc/sys/kernel/random/boot_id` and check it on startup. This id is re-generated with each boot, and if it's different than our stored value, we lock the app.

Reply:
sounds good to me!  The broadcast receiver was the first thing I found about detecting reboot.  As long as `boot_id` exists since API 23 and it can be read without extra permissions, let's do it.